headlamp-k8s/headlamp
# frontend: Add websocket multiplexer

This adds a single websocket connection from frontend to backend and sends lists of messages to the backend. With the help of messages backend creates multiple websocket connection to k8s API and returns back the data.

This solves the issue for limiting of websocket connection in frontend in case of multi cluster setup.

Fixes: #2515 

### Testing

- Run frontend `cd frontend && npm run start #(default is multiplexer enabled)`
- Run backend `cd backend && go run ./cmd -dev -enable-dynamic-clusters`
- Go to pods page http://localhost:3000/c/minikube/pods
- Create a pod in the terminal `kubectl run nginx --image=nginx`
- Everything should be working.
- Smoke test logs and exec inside it.

### TODO

- [x] Frontend UI updated with the data.
- [x] Add reconnect mechanism to keep the frontend and backend websocket connection alive.
- [x] Test coverage up to at least 80% (or less if it makes no sense to test more)
- [x] Stress tested. 
    - [x] Run load tests on multiple clusters to see if it doesn't crash and performs ok.
- [x] Tests run ok with the race detector enabled. `cd backend && go test -race -v -p 1 ./...`
